### ISSUE TITLE
Loss of data during transfer from frontend to backend

### ISSUE THREAD

**Issue by juliosueiras** [UnivJobs/univjobs-main#19]

**Comment by charlesjavelona** [comment]
Here are the error logs:

```python
ValueErroremployer.serializer in update
ErrorCannot assign "OrderedDict([('first_name', 'Yoshi'), ('last_name', 'Mario'), ('mobile', '1295394400'), ('is_profile_completed', True)])": "Business.user" must be a "BaseUser" instance.
```

Looks like error is in **ValueErroremployer.serializer in update**
Will check now.

**Comment by charlesjavelona** [comment]
In addition here is the line error found in **employer/serializer.py in update at line 43**
```python
for user_key in user:
    setattr(request.user, user_key, user[user_key])
request.user.save()

 continue
 **setattr(request, key, args[0][key])**
 request.save()
 eturn request
```

Now moving on to **user/views.py in process_update at line 115**
```python
 def process_update(self, serializer, partial=False):
        if serializer.is_valid():
            # import ipdb
            # ipdb.set_trace()
            **serializer.save(partial=partial)**
            return Response(serializer.data)
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

Then **put at line 159**

```python
@parser_classes((MultiPartParser,))
    def put(self, request):
        import ipdb
        ipdb.set_trace()
        **data = request.data**
        serializer = self.get_user_serializer(self.parse_request_json(data=data))
        return self.process_update(serializer)
```

**Comment by charlesjavelona** [comment]
Did a first test to reproduce bug.

Here is the error below, notice on the right side that captions ValueError, the data that was inputted is not displayed in the OrderedDict. Wonder why that happened?
<img width="1280" alt="screen shot 2016-12-21 at 6 52 53 am" src="https://cloud.githubusercontent.com/assets/7959179/21388427/36466fce-c74a-11e6-88f7-4d9c2b502f9a.png">

**Comment by charlesjavelona** [comment]
Checking the db table **employer_business** it seems that some information was passed.

<img width="1280" alt="screen shot 2016-12-21 at 7 40 11 am" src="https://cloud.githubusercontent.com/assets/7959179/21389704/fcdc362c-c750-11e6-907b-b96d6d68a3eb.png">

**Comment by charlesjavelona** [comment]
There's two questions I have in mind. 
1.) Signup as a user does not work?
2.) Updates on Employer does not work?

**Q1 Answer**
Based on the table query screenshot below, signup works. That eliminates this question.
<img width="1280" alt="screen shot 2016-12-21 at 7 40 11 am" src="https://cloud.githubusercontent.com/assets/7959179/21390607/be578064-c755-11e6-82b6-c7a3bace67df.png">


**Q2 Answer**
Base on this image, PUT does not work. So it seems that somewhere along the receiving the REQUEST call from the front-end, the data is lost. This should give us enough basis to work on.
<img width="1280" alt="screen shot 2016-12-21 at 7 40 11 am" src="https://cloud.githubusercontent.com/assets/7959179/21389704/fcdc362c-c750-11e6-907b-b96d6d68a3eb.png">

Further along, I have the following questions that I would start with:
- Does the request data have all the information from the front-end?
- Could it be during sanitization of data that it was lost?

**Comment by charlesjavelona** [comment]
Further along, I have the following questions that I would start with:
- Does the request data have all the information from the front-end?
- Could it be during sanitization of data that it was lost?

**Comment by charlesjavelona** [comment]
## Update:
Front-end is not a source of problem.

@juliosueiras, below is a good starting point:

<img width="1280" alt="screen shot 2016-12-21 at 10 35 33 am" src="https://cloud.githubusercontent.com/assets/7959179/21395114/434050a4-c769-11e6-9983-3a6fdf251813.png">

**In employers/serializers.py**, args[0] does not contain business. Please refactor somewhere here.

```python
  32     def update(self, request, *args, **kwargs):
   33         import ipdb
   34         ipdb.set_trace()
>> 35         partial = kwargs.pop('partial', False)
   36
   37         #NOTE: Might be the source of errors
   38         for key in args[0]:
   39             #args[0] does not contain business
   40             if key == 'business':
   41                 user = args[0]['business']
   42                 for user_key in user:
   43                     setattr(request.user, user_key, user[user_key])
   44                 request.user.save()
   45                 continue
   46
   47             setattr(request, key, args[0][key])
``

**Comment by charlesjavelona** [comment]
After looking at the data at the backend here is the info on line 
```python
Users/mariacarmenjavelona/Desktop/univjobs-back/univjobs-back/univjobs/user/views.py(104)
```
<img width="1280" alt="screen shot 2016-12-27 at 7 22 18 pm" src="https://cloud.githubusercontent.com/assets/7959179/21511107/4cde191a-cc6a-11e6-953b-9be8807eeaf8.png">

<img width="1280" alt="screen shot 2016-12-27 at 7 22 05 pm" src="https://cloud.githubusercontent.com/assets/7959179/21511111/5af086be-cc6a-11e6-894d-fda50fc427b8.png">

**Comment by charlesjavelona** [comment]
@stemmlerjs, with that said, I have a question, why do we need to get BaseUser info for employers?
Which is first and last name, mobile, is_a_student, ect... they do not get updated.
Would it not make sense to just updated Employer info?

**Comment by stemmlerjs** [comment]
I agree completely that it's unnecessary information to need to supply. I believe we were fdoing it this way based off a requirement from the backend with PUT or PATCH methods. I think PUT required that we get all information because it would perform an "overwrite" of all the fields.

Realistically, we could have just wrote the code to make it perform an "update" instead of overwrite fields that were not included like "firstName" and "last are".

We can do it this way for sure with the new backend.
